compiler/src/iree/compiler/Codegen/Common/ForallToFor.cpp

```diff
@@ -0,0 +1,116 @@
+                                            scf::ForallOp forallOp) {
+  rewriter.setInsertionPoint(forallOp);
+
+  SmallVector<Value> lbs = forallOp.getLowerBound(rewriter);
+  SmallVector<Value> ubs = forallOp.getUpperBound(rewriter);
+  SmallVector<Value> steps = forallOp.getStep(rewriter);
+  SmallVector<Value> iterArgs = forallOp.getOutputs();
+
+  bool unsupportedOperation = false;
+  auto buildBody = [&](OpBuilder &builder, Location loc, ValueRange ivs,
+                       ValueRange args) -> scf::ValueVector {
+    // Inline `scf.forall` body excluding terminating
+    // `scf.forall.in_parallel` op.
+    IRMapping map;
+    map.map(forallOp.getInductionVars(), ivs);
+    map.map(forallOp.getRegionOutArgs(), args);
+    for (auto &op : forallOp.getBody()->without_terminator()) {
+      builder.clone(op, map);
+    }
+
+    // Convert + inline the contents of `scf.forall.in_parallel` terminator.
+    auto terminator = forallOp.getTerminator();
+    SmallVector<Value> yieldedValues;
+    for (auto &yieldOp : terminator.getYieldingOps()) {
+      // Convert tensor.parallel_insert_slice to tensor.insert_slice
+      if (auto parallelInsert =
+              dyn_cast<tensor::ParallelInsertSliceOp>(&yieldOp)) {
+
+        auto insertSlice = builder.create<tensor::InsertSliceOp>(
+            parallelInsert.getLoc(),
+            map.lookupOrDefault(parallelInsert.getSource()),
+            map.lookupOrDefault(parallelInsert.getDest()),
+            llvm::map_to_vector(
+                parallelInsert.getOffsets(),
+                [&](Value v) -> Value { return map.lookupOrDefault(v); }),
+            llvm::map_to_vector(
+                parallelInsert.getSizes(),
+                [&](Value v) -> Value { return map.lookupOrDefault(v); }),
+            llvm::map_to_vector(
+                parallelInsert.getStrides(),
+                [&](Value v) -> Value { return map.lookupOrDefault(v); }),
+            parallelInsert.getStaticOffsets(), parallelInsert.getStaticSizes(),
+            parallelInsert.getStaticStrides());
+        yieldedValues.push_back(insertSlice.getResult());
+      } else {
+        forallOp.emitError("unsupported operation in in_parallel region");
+        unsupportedOperation = true;
+        return args;
+      }
+    }
+
+    return yieldedValues;
+  };
+
+  scf::LoopNest loopNest = scf::buildLoopNest(rewriter, forallOp->getLoc(), lbs,
+                                              ubs, steps, iterArgs, buildBody);
+  if (unsupportedOperation) {
+    return failure();
+  }
+
+  rewriter.replaceOp(forallOp, loopNest.results);
+  return success();
+}
+
+struct ForallToForPass : impl::ForallToForPassBase<ForallToForPass> {
+  using impl::ForallToForPassBase<ForallToForPass>::ForallToForPassBase;
+  void runOnOperation() override {
+    auto funcOp = getOperation();
+
+    IRRewriter rewriter(funcOp->getContext());
+
+    funcOp->walk([&](scf::ForallOp forallOp) {
```

Comment:
Don't rewrite while walking; this can disrupt the use-def chain and sometimes result in crashes. I would rather collect the forall ops from innermost to outermost and start applying rewrites in order.